File: DemoApp/DemoApp/CollectionView/BaseFlowLayoutCollectionViewController.m
//
//  BaseFlowLayoutCollectionViewController.h
//  DemoApp
//
//  Created by Glispa on 22/04/16.
//  Copyright © 2016 Glispa GmbH All rights reserved.
//

#import "BaseFlowLayoutCollectionViewController.h"
#import "StreamAdapterCollectionViewCell.h"
#import "DataUnitManager.h"
#import "DataUnit.h"


@interface BaseFlowLayoutCollectionViewController () <UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout>

@end


@implementation BaseFlowLayoutCollectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.collectionView.showsHorizontalScrollIndicator = NO;
    self.dataSource = [NSMutableArray new];
    [self.collectionView registerNib:[UINib nibWithNibName:NSStringFromClass([StreamAdapterCollectionViewCell class]) bundle:[NSBundle mainBundle]] forCellWithReuseIdentifier:NSStringFromClass([StreamAdapterCollectionViewCell class])];
    [self loadData];
}


#pragma  mark - UICollectionView

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [self.dataSource count];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath {

    StreamAdapterCollectionViewCell *adCell = [collectionView avo_dequeueReusableCellWithReuseIdentifier:NSStringFromClass([StreamAdapterCollectionViewCell class]) forIndexPath:indexPath];
    DataUnit *item = self.dataSource[indexPath.row];

    adCell.tweetNameLabel.text = item.name;
    if (!item.photo) {
        [adCell.tweetImageHeightConstraint setConstant:0];
        adCell.tweetImageView.image = nil;
    } else {
        [adCell.tweetImageHeightConstraint setConstant:(self.collectionView.frame.size.width - 4) / 2];
        adCell.tweetImageView.image = [UIImage imageNamed:item.photo];
    }
    adCell.tweetTextLabel.text = item.specification;

    [adCell layoutIfNeeded];
    return adCell;
}


- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return CGSizeMake((NSInteger) ((self.collectionView.frame.size.width - 4) / 2), (NSInteger) ((self.collectionView.frame.size.width - 4) / 1.1f));
}

- (BOOL)collectionView:(UICollectionView *)collectionView shouldHighlightItemAtIndexPath:(NSIndexPath *)indexPath {
    return NO;
}


#pragma mark - Data

- (void)loadData {
    self.dataSource = [DataUnitManager createDataUnitList:20];
}

@end

File: DemoApp/DemoAppSwift/CollectionView/BaseFlowLayoutCollectionViewController.swift
//
//  BaseFlowLayoutCollectionViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 08.06.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

import Foundation

@objc(BaseFlowLayoutCollectionViewController)
    class BaseFlowLayoutCollectionViewController: BaseCollectionViewController,
                                                  UICollectionViewDelegate,
                                                  UICollectionViewDataSource,
                                                  UICollectionViewDelegateFlowLayout {

    var dataSource: [[DataUnit]] = [[DataUnit]]()

    override func viewDidLoad() {
        super.viewDidLoad()
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.register(UINib.init(nibName: NSStringFromClass(StreamAdapterCollectionViewCell.self), bundle: Bundle.main), forCellWithReuseIdentifier: NSStringFromClass(StreamAdapterCollectionViewCell.self))
    }

    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return dataSource.count
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return dataSource[section].count
    }

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell: StreamAdapterCollectionViewCell = collectionView.avo_dequeueReusableCell(withReuseIdentifier: "StreamAdapterCollectionViewCell", for: indexPath) as! StreamAdapterCollectionViewCell

        let item: DataUnit = self.dataSource[(indexPath as NSIndexPath).section][(indexPath as NSIndexPath).row]
        cell.tweetNameLabel.text = item.name
        if let imageHeightConstraint = cell.tweetImageHeightConstraint {
            if item.photo == nil {
                imageHeightConstraint.constant = 0
            } else {
                imageHeightConstraint.constant = (self.collectionView.frame.size.width - 4) / 2
            }
        }

        cell.tweetImageView.image = UIImage(named: item.photo)

        if let textLabel = cell.tweetTextLabel {
            textLabel.text = item.specification
        }

        cell.layoutIfNeeded()

        return cell
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: Int((collectionView.frame.size.width - 4) / 2.0), height: Int((collectionView.frame.size.width - 4) / 1.1))
    }

    func loadData(_ count: UInt) {
        dataSource = DataUnitManager.createDataUnitList(count, dividedByGroups: 2)
    }
}
